### ISSUE TITLE
SearchIO.QueryResult object does not have an 'accession' attribute all the time when parsing hmmer3-text records

### ISSUE THREAD

**Issue by kblin** [biopython/biopython#1060]
When parsing hmmer3-text records, the `QueryResult` objects returned only have an `accession` attribute when there's an `Accession:` line in the corresponding hmmer file.
This behaviour is at odds with the behaviour on e.g. hmmsearch3-domtab files, where there is always an `accession` attribute on the `QueryResult`, but it may be `-` if not set.

See the zip file below for a reproducer.

[example.zip](https://github.com/biopython/biopython/files/771069/example.zip)

**Comment by kblin** [comment]
If we agree that we want to fix this, the fix should be as trivial as adding a default "accession" attribute to `qresult_attrs` in `_parse_qresult()` for `Bio/SearchIO/HmmerIO/hmmer3_text.py`.

**Comment by peterjc** [comment]
Yes, or ensuring all the parsers leave the object's default value as is if the accession is unknown? Over to Bow to decide.

**Comment by kblin** [comment]
There is no default value.

The relevant snippet is:

```python
            qresult = QueryResult(id=qid, hits=hit_list)
            for attr, value in qresult_attrs.items():
                setattr(qresult, attr, value)
            yield qresult
```

**Comment by bow** [comment]
Thanks for reporting @kblin!

I intentionally left this undefined to better mirror what is represented by the file. In (Bio)python, I'm not sure if we have settled on a unified way to represent this.

We can do at least either of these two:

* Whenever an attribute is missing, set the attribute value to `None`. This means that `hasattr(obj, "attr_name")` will always return `True`.

* Whenever an attribute is missing, we simply don't set it at all. `hasattr(obj, "attr_name")` would then return `False`.

As you see, I chose the latter one throughout `SearchIO`, mainly because I feel it is more intuitive. Whenever an attribute is not present in the actual input, `hasattr` will always return `False`.

Also, having `-` parsed into `None` means that anytime someone has a database sequence with ID `-`, it will be considered as nonexistent. I know this is a super rare case and I'm probably being too pedantic here, but these are the types of errors will take too long to recognize especially since the formats we are parsing do not have any formal specifications.

**Comment by kblin** [comment]
It's just really surprising when code that works just fine with domtab input suddenly fails with an AttributeError just because the file format you're parsing changed.

**Comment by kblin** [comment]
Note that I'm not suggesting to parse `-` into `None`. I'm aware that this is mainly a question of style, but personally I think `if qresult.accession is None or qresult.accession == '-':` reads nicer than `if not hasattr(qresult, 'accession') or qresult.accession == '-':`

**Comment by bow** [comment]
Would this be an acceptable option?

```
def get_hmmer_acc(qresult):
    return None if getattr(qresult, "accession", "-") == "-" else qresult.acc
```

There's the small case of when the ID is actually set to `-`, but I hope prior validation on that is not too much to ask :).

**Comment by bow** [comment]
On second thought, while we're discussing this, would it be an option to add functions like the one above for attributes that are optional to the module as well?

So you'd have `qresult.accession` which may or may not raise an `AttributeError` exception and `qresult.get_accession()` which will always return something.

**Comment by kblin** [comment]
Works for me. I just don't want to worry about code not being portable between different outputs of the same hmmer run.

**Comment by bow** [comment]
@peterjc What do you think about adding a `get_*` method for `accession` for uniform return types? If this is ok, we could also do the same thing for the `description` attributes of `QueryResult` and `Hit` objects. There the, objects without the `description` attribute is has an empty string by default.

**Comment by peterjc** [comment]
If you mean using _private_ get methods as part of a Python property, that seems like a sensible implementation.

(On a related point, the defaults for the ``SeqRecord`` object's ``.id``, ``.name`` and ``.description`` are less than ideal - but changing these is much harder than in ``SearchIO`` as we have more historical usage to worry about for backward compatibility)

**Comment by bow** [comment]
Hmm, I was thinking of making public methods actually.

So anyone using the package can choose, for example, `.accession` to retrieve the closest representation to the file or `.get_accession()` to retrieve what we think is the correct representation (`None` for missing values or whatever is in the file otherwise).

`.get_accession()` always returns something, while `.accession` may or may not raise an `AttributeError` and can be used for `getattr` / `hasattr` calls.

This is only for SearchIO for the moment ~ changing SeqIO means more backward compatibility breaks indeed.

**Comment by peterjc** [comment]
That choice of ways to access the data with a subtly different value seems confusing and un-Pythonic to me. I'd push for just ``.accession`` here only (which might be a plain attribute or a read only property underneath). If the simplest way to do this is to set ``None`` as a default value for missing data, that seems fine to me.

@kblin As someone directly using this and familiar with the data concerned, what do you think?

**Comment by mdehoon** [comment]
FWIW, I agree with @peterjc  's suggestion. I believe it's the most Pythonic solution, and feels most consistent with the principle of least astonishment.

**Comment by kblin** [comment]
Late to the party again, but as I mentioned previously, `if hit.accession is None` in my eyes reads much nicer than `if not hasattr(qresult, 'accession')`, so fine by me. I mainly think having a different selection of attributes for a `QueryResult` depending on what format was parsed is needlessly confusing and likely to cause errors in downstream code.

**Comment by bow** [comment]
Hmm..

Ok then, let's keep using `hit.accession` as a simple attribute then. We could default to using `None` whenever a missing value is encountered. This may be an empty string in some formats, or a `-` (as in hmmer3-domtab), or a value that is altogether not present in the file. SearchIO will have format-specific defaults on what the missing values are, but this may be overridden via an argument to `SearchIO.parse` or `SearchIO.read`.

How does this sound?

**Comment by peterjc** [comment]
Using ``None`` as a default for missing values sounds good. I'm not sure what you have in mind for (extra?) arguments to ``SearchIO.parse`` or ``SearchIO.read``.

**Comment by bow** [comment]
It's just me (probably) being too cautious of a corner case where some records may have an actual ID of `-` in a `hmmer3-domtab` file, I guess. SearchIO would then thinks the ID value is nonexistent and sets it to `None`, while the user may be expecting a `-` value. So maybe giving the user an option to ignore what SearchIO thinks is a blank value is useful.

I don't expect this to happen under any normal use. After all, having `-` as an actual record ID doesn't seem like something most people would do.
